src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsTypedHandler.java

```diff
@@ -0,0 +1,132 @@
+                                  @NotNull Project project,
+                                  @NotNull Editor editor,
+                                  @NotNull PsiFile file,
+                                  @NotNull FileType fileType) {
+        if (LSPIJEditorUtils.isSupportedAbstractFileTypeOrTextMateFile(file)) {
+            if (LSPIJEditorUtils.isQuoteCharacter(file, charTyped) && handleNestedQuote(file, editor, charTyped)) {
+                return Result.STOP;
+            }
+            // TODO: Get statement terminator character(s) from client config?
+            else if ((';' == charTyped) && handleStatementTerminator(file, editor)) {
+                return Result.STOP;
+            }
+        }
+
+        return Result.CONTINUE;
+    }
+
+    private boolean handleNestedQuote(@NotNull PsiFile file,
+                                      @NotNull Editor editor,
+                                      char quote) {
+        Document document = editor.getDocument();
+        CharSequence documentChars = document.getCharsSequence();
+        int offset = editor.getCaretModel().getOffset();
+
+        // Try to find a string literal that contains the current offset
+        String stringLiteral = getStringLiteral(file, documentChars, offset);
+        if (StringUtil.isNotEmpty(stringLiteral)) {
+            char stringLiteralQuote = stringLiteral.charAt(0);
+            // Single quote within double-quoted string or vice-versa
+            // Escaped single quote within single-quoted string or vice-versa
+            if ((stringLiteralQuote != quote) || ((offset > 0) && (documentChars.charAt(offset - 1) == '\\'))) {
+                document.insertString(offset, String.valueOf(quote));
+                editor.getCaretModel().moveToOffset(offset + 1);
+                return true;
+            }
+        }
+        // If the same quote character used for the string is typed immediately after an escape and before the string
+        // literal's closing quote character, no string literal will be found, so just insert it and advance the caret
+        else if ((offset > 0) && (documentChars.charAt(offset - 1) == '\\')) {
+            document.insertString(offset, String.valueOf(quote));
+            editor.getCaretModel().moveToOffset(offset + 1);
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean handleStatementTerminator(@NotNull PsiFile file,
+                                              @NotNull Editor editor) {
+        Document document = editor.getDocument();
+        CaretModel caretModel = editor.getCaretModel();
+        int offset = caretModel.getOffset();
+        if (offset == document.getTextLength()) {
+            return false;
+        }
+
+        CharSequence documentChars = document.getCharsSequence();
+        char nextCharacter = documentChars.charAt(offset);
+        if ((nextCharacter == ';') && !inStringLiteral(file, documentChars, offset)) {
```

Comment:
If typing a statement terminator outside of a string literal and the next character is that statement terinator, just advance the caret. This is the exact same behavior as in natively-supported languages like Java.

Note that in my own implementation, I also confirm that it's not inside a comment, but I also have lightweight ways to check whether or not the caret is inside of a JavaScript/TypeScript or CSS comment because I know those languages' comment patterns. We don't have that same information generically for LSP-supported languages, though, so I didn't include that check in this port.